### PR TITLE
JSSpecCompiler: Work on the `-xspec` parser (and a test for it!)

### PR DESCRIPTION
**Pull request by DanShaders**
I also stopped trying to make diff ~250 lines since I doubt it significantly influences review time. But in case I'm wrong I'll happily continue my practice of submitting moderately small PRs.

### CODE REVIEW

**Comment by DanShaders** [comment]
Actually, I lied in the 3th commit description -- we can't parse any. But with an additional 2 line fix, the count goes up to 17 (and their AST look correct on the first glance)!

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * Copyright (c) 2024, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DiagnosticEngine.h"
+
+namespace JSSpecCompiler {
+
+thread_local DiagnosticEngine* DiagnosticEngine::m_instance = nullptr;
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.cpp:11]: Does this really need to be thread_local? Or are you intending to multi-thread this thing and have different diagnostics per thread?

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Function.h [hunk 2]
```diff
@@ -11,15 +11,20 @@
...
 
 struct TranslationUnit {
+    TranslationUnit(StringView filename);
+    ~TranslationUnit();
+
     void adopt_declaration(NonnullRefPtr<FunctionDeclaration>&& declaration);
     FunctionDefinitionRef adopt_function(NonnullRefPtr<FunctionDefinition>&& definition);
 
     StringView filename;
+    DiagnosticEngine diagnostic_engine;
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Function.h:27]: If the translation unit owns the Diagnostic engine, why does it need to be globally accessible? Can't the parser just do translation_unit->diag().fatal_error() or some such?

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/ParseError.h [hunk 3]
```diff
@@ -24,6 +26,7 @@ class ParseError : public RefCounted<ParseError> {
     static NonnullRefPtr<ParseError> create(ErrorOr<String> message, XML::Node const* node);
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/Parser/ParseError.h:26]: perhaps we should be asking the Diagnostic engine to create our errors for us? Then we wouldn't need the "Current file" concept.

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    static Location file_scope();
+    static Location from_xml_offset(XML::Offset offset);
+    static Location from_xml_offset(XML::Offset offset, NonnullRefPtr<LogicalLocation> logical_location);
+};
+
+class DiagnosticEngine {
+public:
+    DiagnosticEngine(StringView filename);
+
+    static DiagnosticEngine& instance();
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h:39]: we use the() for singletons, if this really needs to be one.

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+struct LogicalLocation : RefCounted<LogicalLocation> {
+    StringView filename;
+    String section;
+    String step;
+};
+
+struct Location {
+    NonnullRefPtr<LogicalLocation> logical_location;
+    size_t line = 0, column = 0;
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h:27]: one declaration per line, please

File: Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        String message;
+
+        Vector<Diagnostic> notes;
+
+        bool operator<(Diagnostic const& other) const;
+
+        void format_into(StringBuilder& builder, UseColor) const;
+    };
+
+    friend Location;
```
Comment (ADKaster) [inline Meta/Lagom/Tools/CodeGenerators/JSSpecCompiler/DiagnosticEngine.h:95]: If the only thing we need from the DiagnosticEngine is the current file, I would prefer a public "filename()" getter to friendship.

**Review by ADKaster — CHANGES_REQUESTED** [review]
Seems overall good progress, but I'm skeptical on the architecture of the DiagnosticEngine. 

- Does it really need to be a singleton? It's owned by the current TU, and needs to be scoped as such.
- The with() function to change the instance seems awkward to use


Is the DiagnosticEngine really the right level to be creating a global state? I think clang has like a SourceManager or some such, which at least makes it obvious that you're reaching into per-TU data.

**Comment by DanShaders** [comment]
@ADKaster Questions in comments are not applicable anymore :) I've redesigned the whole thing so that `DiagnosticEngine` is now fully owned by TU and `Location`s are only created by `SpecificationParsingContext` which I pass around as a second argument.

**Review by ADKaster — APPROVED** [review]

**Comment by ADKaster** [comment]
Very nice! I like the context-based approach better than the global  that gets swapped out all the time approach much better.

**Pull request merged**